### PR TITLE
Enable rich version of Brave Search autocomplete

### PR DESCRIPTION
**Pull request by sangwoo108**
* In order to provides users richer information about keywords, adds `rich=true` param to Brave Search's suggestion url

* Consequently, we need to handle the new format of data returned. This requires patching upstream's logic to parse suggestion.

Resolves https://github.com/brave/brave-browser/issues/29997

image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
* Enable settings > Search Engine > Improve Search Suggestion
* Type something
* See if there're entities with image and description

### CODE REVIEW

**Comment by simonhong** [comment]
Sometimes result icon is empty. Can we show default icon instead in this case?
image

**Comment by sangwoo108** [comment]
> Sometimes result icon is empty. Can we show default icon instead in this case? image

Seems to be the image_url is svg in that case or empty. I'm not sure what should be shown instead. Magnifier seems to be a little bit off to me. How about handling it from a follow up PR? 

@aguscruiz , have any ideas? We have suggestion entity, which is equivalent to what is used in Brave Search's search box. In case we can't render image of suggestion entity, what should be shown?

**Comment by simonhong** [comment]
I think magnifier icon seems fine if icon is not available like below.
`email login` doesn't have icon and magnifier icon is set instead while others are empty.
All these three items are from search suggestions.
IMO, it would be better to handling it in this PR as it seems like bug/regressions if it's empty.
Screenshot 2024-05-30 at 10 25 37 AM

**Comment by sangwoo108** [comment]
Okay, as the entity image's visibility is decided with the image url, added code to filter .svg.

File: chromium_src/components/omnibox/browser/search_suggestion_parser.cc [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+// static
+bool SearchSuggestionParser::ParseSuggestResults(
+    const base::Value::List& root_list,
+    const AutocompleteInput& input,
+    const AutocompleteSchemeClassifier& scheme_classifier,
+    int default_result_relevance,
+    bool is_keyword_result,
+    Results* results,
+    bool is_brave_rich_suggestion) {
+  if (is_brave_rich_suggestion) {
```
Comment (simonhong) [inline chromium_src/components/omnibox/browser/search_suggestion_parser.cc:29]: Q) we only hit here when `Improve search suggestions` option is on?

File: components/search_engines/brave_prepopulated_engines.cc [hunk 2]
```diff
@@ -194,7 +194,7 @@ const PrepopulatedEngine brave_search = MakeBravePrepopulatedEngine(
     "desktop",
 #endif
     "UTF-8",
-    "https://search.brave.com/api/suggest?q={searchTerms}",
+    "https://search.brave.com/api/suggest?q={searchTerms}&rich=true",
```
Comment (simonhong) [inline components/search_engines/brave_prepopulated_engines.cc:197]: I think `kBraveCurrentDataVersion` should be updated.

Comment (sangwoo108) [reply components/search_engines/brave_prepopulated_engines.cc:197]: Nice find! Thanks

File: chromium_src/components/omnibox/browser/search_suggestion_parser.cc (hunk 1 above)
Comment (sangwoo108) [reply chromium_src/components/omnibox/browser/search_suggestion_parser.cc:29]: Yes. I think SearchProvider would be disabled when the option is off.

**Review by simonhong — APPROVED** [review]
++ 👍🏼 
it would be great to add some comment how to test this such as turn on `Improve search suggestions` in `Test Plan`

File: chromium_src/components/omnibox/browser/search_suggestion_parser.cc [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          /*match_contents_prefix*/ {},
+          /*annotation*/ annotation, std::move(entity_info),
+          /*deletion_url*/ {}, is_keyword_result, omnibox::NAV_INTENT_NONE,
+          /*relevance*/ -1,
+          /*relevance_from_server*/ false, false, false,
+          base::CollapseWhitespace(input_text, false));
+
+      results->suggest_results.push_back(std::move(result));
+    }
+    return true;
```
Comment (goodov) [inline chromium_src/components/omnibox/browser/search_suggestion_parser.cc:133]: would be great to have it as a separate function with its own unit tests.

**Comment by aguscruiz** [comment]
I think the magnifier icon is fine if we can't load it, as long as we make it look like the other magnifier icons in the omnibar.

one thing though, from this screenshot, I noticed the small images are vertical rectangles, they should all be squared.
image

----------------------

Here, added a rich result here on the Figma library:
https://www.figma.com/design/H11ZOl6JMYbCTW4ZJXqR5V/%F0%9F%A6%81-Browser?node-id=1386-11575&t=ZzK9Ybhkb0CHB4mq-1

image

File: chromium_src/components/omnibox/browser/search_suggestion_parser.cc (hunk 3 above)
Comment (sangwoo108) [reply chromium_src/components/omnibox/browser/search_suggestion_parser.cc]: Good call

**Comment by sangwoo108** [comment]
@goodov Could you take another look?

**Review by goodov — APPROVED** [review]

**Review by ecnmst — APPROVED** [review]

**Pull request merged**